File: src/core/resource.ts
// Canonical Core — resource addressing.
//
// A resource is named "<ownerId>:<path>". The owner is the identity to the left
// of the first colon (identity ids are base64url and contain no colon). Paths
// may end in "/*" to denote a subtree; "*" denotes everything.

export function makeResource(ownerId: string, path: string): string {
  return ownerId + ":" + path;
}

export function resourceOwner(resource: string): string {
  const i = resource.indexOf(":");
  return i < 0 ? resource : resource.slice(0, i);
}

export function resourcePath(resource: string): string {
  const i = resource.indexOf(":");
  return i < 0 ? "" : resource.slice(i + 1);
}

export function pathSubsumes(parent: string, child: string): boolean {
  if (parent === child) return true;
  if (parent === "*") return true;
  if (parent.endsWith("/*")) return child.startsWith(parent.slice(0, -1));
  return false;
}

/** True when `parent` covers `child`: same owner and the path subsumes. */
export function resourceSubsumes(parent: string, child: string): boolean {
  return resourceOwner(parent) === resourceOwner(child) && pathSubsumes(resourcePath(parent), resourcePath(child));
}

export function actionSubsumes(parent: string, child: string): boolean {
  return parent === "*" || parent === child;
}

File: crates/canonical-core/src/resource.rs
//! Resource addressing: "<ownerId>:<path>". The owner is everything before the
//! first colon (identity ids are base64url and contain no colon). A path may end
//! in "/*" for a subtree; "*" is everything.

pub fn make_resource(owner_id: &str, path: &str) -> String {
    format!("{owner_id}:{path}")
}

pub fn resource_owner(resource: &str) -> &str {
    match resource.find(':') {
        Some(i) => &resource[..i],
        None => resource,
    }
}

pub fn resource_path(resource: &str) -> &str {
    match resource.find(':') {
        Some(i) => &resource[i + 1..],
        None => "",
    }
}

pub fn path_subsumes(parent: &str, child: &str) -> bool {
    if parent == child || parent == "*" {
        return true;
    }
    if let Some(prefix) = parent.strip_suffix("/*") {
        return child.starts_with(prefix) && child.len() > prefix.len();
    }
    false
}

pub fn resource_subsumes(parent: &str, child: &str) -> bool {
    resource_owner(parent) == resource_owner(child)
        && path_subsumes(resource_path(parent), resource_path(child))
}

pub fn action_subsumes(parent: &str, child: &str) -> bool {
    parent == "*" || parent == child
}
